packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
+  // `mostRecentEvent`.
+  //
+  // The returned value is never null.
+  _MouseState _guaranteeMouseState(int device, PointerEvent mostRecentEvent) {
+    final _MouseState currentState = _mouseStates[device];
+    if (currentState == null) {
+      _addMouseDevice(device, mostRecentEvent);
+    }
+    final _MouseState result = currentState ?? _mouseStates[device];
+    assert(result != null);
+    return result;
+  }
 
-  /// Track an annotation so that if the mouse enters it, we send it events.
-  ///
-  /// This is typically called when the [AnnotatedRegion] containing this
-  /// annotation has been added to the layer tree.
-  void attachAnnotation(MouseTrackerAnnotation annotation) {
-    _trackedAnnotations[annotation] = _TrackedAnnotation(annotation);
-    // Schedule a check so that we test this new annotation to see if any mouse
-    // is currently inside its region. It has to happen after the frame is
-    // complete so that the annotation layer has been added before the check.
-    if (mouseIsConnected) {
-      _scheduleMousePositionCheck();
+  // The collection of annotations that are currently being tracked.
+  // It is operated by [attachAnnotation] and [detachAnnotation].
+  final Set<MouseTrackerAnnotation> _trackedAnnotations = <MouseTrackerAnnotation>{};
+  bool get _hasAttachedAnnotations => _trackedAnnotations.isNotEmpty;
+
+  void _addMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(!_mouseStates.containsKey(device));
+    _mouseStates[device] = _MouseState(mostRecentEvent: event);
+    // Schedule a check to enter annotations that might contain this pointer.
+    _checkDeviceUpdates(device: device);
+    if (mouseIsConnected != wasConnected) {
+      notifyListeners();
     }
   }
 
-  /// Stops tracking an annotation, indicating that it has been removed from the
-  /// layer tree.
-  ///
-  /// An assertion error will be thrown if the associated layer is not removed
-  /// and receives another mouse hit.
-  void detachAnnotation(MouseTrackerAnnotation annotation) {
-    final _TrackedAnnotation trackedAnnotation = _findAnnotation(annotation);
-    for (int deviceId in trackedAnnotation.activeDevices) {
-      if (annotation.onExit != null) {
-        final PointerEvent event = _lastMouseEvent[deviceId] ?? _pendingRemovals[deviceId];
-        assert(event != null);
-        annotation.onExit(PointerExitEvent.fromMouseEvent(event));
+  void _removeMouseDevice(int device, PointerEvent event) {
+    final bool wasConnected = mouseIsConnected;
+    assert(_mouseStates.containsKey(device));
+    final _MouseState disconnectedMouseState = _mouseStates.remove(device);
+    disconnectedMouseState.mostRecentEvent = event;
+    // Schedule a check to exit annotations that used to contain this pointer.
+    _checkDeviceUpdates(
+      device: device,
+      disconnectedMouseState: disconnectedMouseState,
```

Comment:
Why do you need to pass this in separately (seems like it complicates the logic in this method as well)? Could you add a boolean "disconnected" to MouseState and flip it to true here instead of removing the mouse state from _mouseStates. And then after _checkDeviceUpdates comes back, you remove it from the list?

Reply:
I'm concerned that keeping disconnected states in `_mouseStates` will affect `mouseIsConnected`. If the callback called by `_checkDeviceUpdates` checks `mouseIsConnected`, it'll get a `true` even when the last mouse has been disconnected.

Reply:
I see. Makes sense.